File: packages/quran-engine-kotlin/src/test/kotlin/com/quranengine/ParityTest.kt
package com.quranengine

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInstance

/**
 * The modules beyond the core: the mushaf / word-by-word / Ask AI batch, and the surah outlines,
 * alphabet reference and qiraat comparison that followed it.
 *
 * Mirrors the JS `test/parity.test.js`, the Python `tests/test_parity.py`, the Swift
 * `ParityTests.swift`, the Rust `tests/parity.rs` and the Go `parity_test.go` case for case, so a
 * divergence between the ports shows up as a failing test rather than as a surprise in an app.
 */
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class ParityTest {

    private lateinit var engine: Engine

    @BeforeAll
    fun setup() {
        val dataDir = Engine.findDefaultDataDir()
        assertNotNull(dataDir, "Could not locate data/quran.json by walking up from the working dir")
        engine = Engine.load(
            dataDir,
            loadSurahInfo = true,
            loadMushaf = true,
            loadQiraatTajweed = true,
            loadWordByWord = true,
            loadSimilarAyahs = true,
            loadQiraat = true,
        )
    }

    // ---- mushaf -------------------------------------------------------------------------

    @Test
    fun `twenty riwayat, eight with text`() {
        assertEquals(20, engine.mushaf.riwayat().size)
        assertEquals(8, engine.mushaf.riwayatWithText().size)
        assertEquals(604, engine.mushaf.totalPages())
        assertEquals("Nafi", engine.mushaf.riwayah("warsh")?.imam)
        assertEquals(false, engine.mushaf.riwayah("hisham")?.textIncluded)
    }

    @Test
    fun `every riwayah has a facsimile and a full page table`() {
        for (entry in engine.mushaf.riwayat()) {
            assertTrue(entry.pdf.startsWith("pdfs/"), entry.riwayah)
            assertTrue(entry.pdf.endsWith(".pdf.xz"), entry.riwayah)
            assertTrue(entry.pdfBytes > 100_000, entry.riwayah)
            // Al-Fatihah opens page 1 and an-Nas closes page 604 in every print of the set.
            assertEquals(1, engine.mushaf.page(1, 1, entry.riwayah), entry.riwayah)
            assertEquals(604, engine.mushaf.page(114, 6, entry.riwayah), entry.riwayah)
        }
    }

    @Test
    fun `pages resolve back to their ayahs`() {
        assertEquals(42, engine.mushaf.page(2, 255, "hafs"))
        assertTrue(engine.mushaf.ayahsOnPage(42, "hafs").any { it.surah == 2 && it.ayah == 255 })
        assertEquals(VerseRef(1, 1), engine.mushaf.firstAyahOfPage(1, "hafs"))
    }

    @Test
    fun `line tables ship exactly where the text does`() {
        for (entry in engine.mushaf.riwayat()) {
            val breaks = engine.mushaf.lineBreaks(1, 1, entry.riwayah)
            assertEquals(entry.textIncluded, breaks != null, entry.riwayah)
        }
    }

    // ---- riwayah tajweed -----------------------------------------------------------------

    @Test
    fun `seven verified packs`() {
        assertEquals(
            listOf("buzzi", "duri", "qaloon", "qunbul", "shubah", "susi", "warsh"),
            engine.qiraatTajweed.available(),
        )
    }

    @Test
    fun `legend carries its explanation`() {
        val legend = engine.qiraatTajweed.legend("warsh")
        assertTrue(legend.size >= 3)
        for (entry in legend) {
            assertEquals(1, entry.code.length)
            assertTrue(entry.rule.isNotEmpty() && entry.arabic.isNotEmpty() && entry.english.isNotEmpty())
            assertTrue(entry.short.isNotEmpty(), "no description for ${entry.rule}")
        }
    }

    @Test
    fun `word rules name real legend codes`() {
        val rules = engine.qiraatTajweed.wordRules(2, 3, "warsh")
        assertTrue(rules.isNotEmpty())
        val codes = engine.qiraatTajweed.legend("warsh").map { it.code }
        for (rule in rules) {
            assertTrue(rule.code in codes, rule.code)
            assertEquals(rule.firstLetter < 0, rule.wholeWord)
            assertTrue(rule.word >= 1)
        }
    }

    @Test
    fun `khilaf markers`() {
        assertTrue(engine.qiraatTajweed.hasKhilaf(2, 253, "warsh"))
        assertTrue(!engine.qiraatTajweed.hasKhilaf(2, 254, "warsh"))
    }

    // ---- word by word ---------------------------------------------------------------------

    @Test
    fun `both layers aligned to the ayah's own tokens`() {
        val words = engine.wordByWord.words(112, 1)
        assertEquals(listOf("qul", "huwa", "l-lahu", "aḥadun"), words.map { it.transliteration })
        assertEquals("Say", words[0].english)
        val firstToken = engine.quran.ayah(112, 1)!!.textArabic.trim().split(Regex("(?U)\\s+"))[0]
        assertEquals(firstToken, words[0].arabic)
    }

    @Test
    fun `every ayah's arrays match its token count`() {
        for (surah in engine.quran.all()) {
            for (ayah in surah.ayahs) {
                val tokens = ayah.textArabic.trim().split(Regex("(?U)\\s+")).size
                assertEquals(
                    tokens, engine.wordByWord.glosses(surah.id, ayah.id)?.size,
                    "${surah.id}:${ayah.id} english",
                )
                assertEquals(
                    tokens, engine.wordByWord.transliterations(surah.id, ayah.id)?.size,
                    "${surah.id}:${ayah.id} transliteration",
                )
            }
        }
    }

    @Test
    fun `gloss search finds the word`() {
        val hits = engine.wordByWord.find("the Ever-Living", limit = 5)
        assertTrue(hits.any { it.surah == 2 && it.ayah == 255 })
        assertTrue(hits.all { it.transliteration.isNotEmpty() })
    }

    // ---- similar ayahs, themes, lessons ------------------------------------------------------

    @Test
    fun `similar ayahs`() {
        val matches = engine.similarAyahs.matches(2, 255)
        val found = matches.firstOrNull { it.surah == 3 && it.ayah == 2 }
        assertNotNull(found, "3:2 is a match")
        assertTrue(found!!.verified)
        assertTrue(engine.similarAyahs.has(2, 255))
        assertTrue(engine.similarAyahs.count() > 5000)
    }

    @Test
    fun `themes index both ways`() {
        assertTrue(engine.themes.all().size >= 300)
        assertTrue(engine.themes.topic("tawheed")!!.ayahs.contains("2:255"))
        assertTrue(engine.themes.topicsFor(2, 255).any { it.id == "tawheed" })
        assertTrue(engine.themes.domains().size >= 2)
    }

    @Test
    fun `lessons walk in course order`() {
        val lessons = engine.tajweedLessons.allLessons()
        assertTrue(lessons.size >= 30)
        assertNull(engine.tajweedLessons.previous(lessons[0].id))
        assertEquals(lessons[1].id, engine.tajweedLessons.next(lessons[0].id)?.id)
        assertNotNull(engine.tajweedLessons.chapterOf(lessons[0].id))
    }

    // ---- meaning search ---------------------------------------------------------------------

    @Test
    fun `MaxSim ranks by meaning`() {
        // A toy embedder: enough to prove the scoring, without shipping a model.
        val vectors = mapOf(
            "patience" to floatArrayOf(1.0f, 0.0f, 0.0f),
            "hardship" to floatArrayOf(0.9f, 0.1f, 0.0f),
            "sabr" to floatArrayOf(0.95f, 0.05f, 0.0f),
            "steadfast" to floatArrayOf(0.9f, 0.05f, 0.0f),
            "dawn" to floatArrayOf(0.0f, 1.0f, 0.0f),
            "prayer" to floatArrayOf(0.0f, 0.95f, 0.0f),
        )
        val semantic = Semantic(embed = { vectors[it] }).index(
            listOf(
                Semantic.Document("sabr", "sabr and steadfast endurance"),
                Semantic.Document("fajr", "prayer at dawn"),
            )
        )
        val hits = semantic.search("patience hardship")
        assertEquals("sabr", hits[0].id)
        assertTrue(hits[0].score > hits[1].score)
    }

    // ---- ask AI ------------------------------------------------------------------------------

    @Test
    fun `named verse is the subject`() {
        val passages = engine.askAI.retrieve("explain ayat al-kursi")
        assertEquals("2:255", passages[0].reference)
        assertTrue(passages[0].isSubject)
    }

    @Test
    fun `named surah answers with its background`() {
        val passages = engine.askAI.retrieve("what is surah al-kahf about")
        assertEquals("Surah Al-Kahf", passages[0].reference)
        assertEquals(AskAI.Kind.SURAH, passages[0].kind)
    }

    @Test
    fun `keyword lane is weighted`() {
        val passages = engine.askAI.retrieve("what does the Quran say about patience in hardship")
        assertTrue(passages.any { it.reference == "2:153" })
        for (passage in passages) {
            if (passage.kind == AskAI.Kind.AYAH) {
                assertNotNull(engine.quran.ayah(passage.surah!!, passage.ayah!!))
            }
        }
    }

    @Test
    fun `bare follow-up uses the previous question`() {
        val carried = engine.askAI.retrieve("tell me about 2:153")
        assertTrue(engine.askAI.retrieve("why?").isEmpty())
        val withContext = engine.askAI.retrieve(
            "why?",
            previousQuestion = "tell me about 2:153",
            carried = carried,
        )
        assertTrue(withContext.any { it.reference == "2:153" })
    }

    @Test
    fun `prompt shape`() {
        val passages = engine.askAI.retrieve("explain 2:153")
        val (instructions, prompt) = AskAI.chatPrompt("explain 2:153", passages)
        assertTrue(instructions.contains("Never issue a religious ruling"))
        assertTrue(prompt.contains("SUBJECT OF THE QUESTION [2:153]"))
        assertTrue(prompt.trimEnd().endsWith("QUESTION: explain 2:153"))
        assertTrue("what" in AskAI.QUESTION_WORDS)
    }

    // ---- surah sections ----------------------------------------------------------------

    @Test
    fun `111 surahs carry an outline, and it reads as a chain`() {
        assertEquals(111, engine.surahSections.count())
        assertTrue(engine.surahSections.overview(1).length > 20)
        assertTrue(!engine.surahSections.hasSections(1))

        // Hud opens with a broad passage and the sections inside it - an ayah has a chain, not a row.
        assertEquals(
            listOf("Doctrine facts", "Calling to Allah"),
            engine.surahSections.sectionsFor(11, 3).map { it.english },
        )
        assertEquals("Calling to Allah", engine.surahSections.sectionFor(11, 3)?.english)
    }

    @Test
    fun `the outline rebuilds the nesting the flat list encodes`() {
        val roots = engine.surahSections.outline(11)
        assertTrue(roots.size >= 2)
        assertEquals("Doctrine facts", roots[0].section.english)
        assertEquals(5, roots[0].children.size)
        assertTrue(roots[0].children.all {
            it.section.from >= roots[0].section.from && it.section.to <= roots[0].section.to
        })
    }

    @Test
    fun `every section range is inside its surah`() {
        for (surah in engine.quran.all()) {
            for (section in engine.surahSections.sections(surah.id)) {
                assertTrue(
                    section.from >= 1 && section.to <= surah.numberOfAyahs && section.from <= section.to,
                    "${surah.id}: ${section.from}-${section.to}",
                )
                assertTrue(section.english.isNotEmpty() && section.arabic.isNotEmpty())
            }
        }
        val nuh = engine.surahSections.search("Story of Nuh")
        assertTrue(nuh.size >= 2)
        assertTrue(nuh.any { (surah, _) -> surah == 11 })
    }

    // ---- arabic alphabet ----------------------------------------------------------------

    @Test
    fun `28 letters, each with a tajweed weight the catalogue explains`() {
        val letters = engine.alphabet.letters()
        assertEquals(28, letters.size)
        val weights = engine.alphabet.weightDescriptions()
        for (letter in letters) {
            assertTrue(letter.letter.isNotEmpty() && letter.name.isNotEmpty())
            val weight = letter.weight
            assertNotNull(weight, "${letter.letter} has no weight")
            assertNotNull(weights[weight], "no description for weight $weight")
        }
        // Alif is the one letter with no weight of its own - the fact the whole field exists for.
        assertEquals("followsPrevious", engine.alphabet.weight("ا"))
        assertEquals("heavy", engine.alphabet.weight("ص"))
        assertEquals("light", engine.alphabet.weight("س"))
    }

    @Test
    fun `a letter resolves from any of its joining forms`() {
        assertEquals("Saad", engine.alphabet.letter("ـصـ")?.transliteration)
        assertEquals("ا", engine.alphabet.letterById(1)?.letter)
        assertNull(engine.alphabet.letter("nope"))
    }

    @Test
    fun `tashkeel, numerals and the waqf signs`() {
        assertTrue(engine.alphabet.tashkeel().size >= 8)
        assertEquals(11, engine.alphabet.numbers().size)
        assertEquals("Make Sujood", engine.alphabet.stoppingSign("۩")?.title)
        assertTrue(engine.alphabet.heavyLetters().size >= 7)
    }

    // ---- qiraat comparison ---------------------------------------------------------------

    @Test
    fun `only the published readings, hafs included`() {
        assertEquals(
            listOf("buzzi", "duri", "hafs", "qaloon", "qunbul", "shubah", "susi", "warsh"),
            engine.qiraatComparison.available(),
        )
    }

    @Test
    fun `a reading against itself is entirely identical`() {
        val same = engine.qiraatComparison.compareSurah(2, "hafs")
        assertEquals(same.words, same.identical)
        assertEquals(0, same.sameSkeleton)
        assertEquals(0, same.different)
        assertEquals(0, same.added)
        assertEquals(0, same.dropped)
        assertEquals(100.0, same.identicalPercent)
    }

    @Test
    fun `the buckets add up, and Shubah is nearer to Hafs than Warsh is`() {
        val warsh = engine.qiraatComparison.compareSurah(2, "warsh")
        val shubah = engine.qiraatComparison.compareSurah(2, "shubah")
        for (totals in listOf(warsh, shubah)) {
            assertEquals(
                totals.words,
                totals.identical + totals.sameSkeleton + totals.different + totals.dropped,
            )
        }
        // Shu'bah and Hafs are the two narrators of Asim; Warsh is a different imam entirely.
        assertTrue(shubah.identicalPercent > warsh.identicalPercent)
        assertTrue(shubah.identicalPercent > 95)
    }

    @Test
    fun `differences are the non-identical rows, in reading order`() {
        val rows = engine.qiraatComparison.differences(2, "warsh", limit = 20)
        assertTrue(rows.isNotEmpty())
        assertTrue(rows.all { it.base != it.other })
        assertEquals(rows.map { it.position }.sorted(), rows.map { it.position })
    }

    @Test
    fun `word streams follow the reading's own verse count`() {
        // Warsh reads al-Baqarah as 285 verses to Hafs' 286, so pairing by ayah id would compare
        // different verses from that point on; the comparison walks the surah's words instead.
        assertEquals(285, engine.quran.numberOfAyahsInQiraah(2, "warsh"))
        assertEquals(285, engine.quran.qiraahVerses(2, "warsh").size)
        assertTrue(engine.qiraatComparison.words(2, "warsh").size > 6000)
    }
}

File: packages/quran-engine-dart/test/parity_test.dart
// The modules beyond the core: the mushaf / word-by-word / Ask AI batch, and the
// surah outlines, alphabet reference and qiraat comparison that followed it.
//
// Mirrors the JS `test/parity.test.js`, the Python `tests/test_parity.py`, the
// Swift `ParityTests.swift`, the Rust `tests/parity.rs`, the Go `parity_test.go`
// and the Kotlin `ParityTest.kt` case for case, so a divergence between the
// ports shows up as a failing test rather than as a surprise in an app.

import 'package:quran_engine/quran_engine.dart';
import 'package:test/test.dart';

final RegExp _whitespace = RegExp(r'\s+');

void main() {
  late Engine engine;

  setUpAll(() async {
    engine = await Engine.load(
      loadMushaf: true,
      loadQiraatTajweed: true,
      loadWordByWord: true,
      loadSimilarAyahs: true,
      loadQiraat: true,
    );
  });

  // ---- mushaf -------------------------------------------------------------

  group('mushaf', () {
    test('twenty riwayat, eight with text', () {
      expect(engine.mushaf.riwayat().length, 20);
      expect(engine.mushaf.riwayatWithText().length, 8);
      expect(engine.mushaf.totalPages(), 604);
      expect(engine.mushaf.riwayah('warsh')!.imam, 'Nafi');
      expect(engine.mushaf.riwayah('hisham')!.textIncluded, isFalse);
    });

    test('every riwayah has a facsimile and a full page table', () {
      for (final entry in engine.mushaf.riwayat()) {
        expect(entry.pdf, startsWith('pdfs/'), reason: entry.riwayah);
        expect(entry.pdf, endsWith('.pdf.xz'), reason: entry.riwayah);
        expect(entry.pdfBytes, greaterThan(100000), reason: entry.riwayah);
        // Al-Fatihah opens page 1 and an-Nas closes page 604 in every print.
        expect(engine.mushaf.page(1, 1, entry.riwayah), 1, reason: entry.riwayah);
        expect(engine.mushaf.page(114, 6, entry.riwayah), 604, reason: entry.riwayah);
      }
    });

    test('pages resolve back to their ayahs', () {
      expect(engine.mushaf.page(2, 255, 'hafs'), 42);
      expect(
        engine.mushaf.ayahsOnPage(42, 'hafs').any((r) => r.surah == 2 && r.ayah == 255),
        isTrue,
      );
      expect(engine.mushaf.firstAyahOfPage(1, 'hafs'), const VerseRef(1, 1));
    });

    test('line tables ship exactly where the text does', () {
      for (final entry in engine.mushaf.riwayat()) {
        final breaks = engine.mushaf.lineBreaks(1, 1, entry.riwayah);
        expect(breaks != null, entry.textIncluded, reason: entry.riwayah);
      }
    });
  });

  // ---- riwayah tajweed ----------------------------------------------------

  group('riwayah tajweed', () {
    test('seven verified packs', () {
      expect(engine.qiraatTajweed.available(),
          ['buzzi', 'duri', 'qaloon', 'qunbul', 'shubah', 'susi', 'warsh']);
    });

    test('legend carries its explanation', () {
      final legend = engine.qiraatTajweed.legend('warsh');
      expect(legend.length, greaterThanOrEqualTo(3));
      for (final entry in legend) {
        expect(entry.code.length, 1);
        expect(entry.rule, isNotEmpty);
        expect(entry.arabic, isNotEmpty);
        expect(entry.english, isNotEmpty);
        expect(entry.short, isNotEmpty, reason: 'no description for ${entry.rule}');
      }
    });

    test('word rules name real legend codes', () {
      final rules = engine.qiraatTajweed.wordRules(2, 3, 'warsh');
      expect(rules, isNotEmpty);
      final codes = engine.qiraatTajweed.legend('warsh').map((e) => e.code).toList();
      for (final rule in rules) {
        expect(codes, contains(rule.code));
        expect(rule.wholeWord, rule.firstLetter < 0);
        expect(rule.word, greaterThanOrEqualTo(1));
      }
    });

    test('khilaf markers', () {
      expect(engine.qiraatTajweed.hasKhilaf(2, 253, 'warsh'), isTrue);
      expect(engine.qiraatTajweed.hasKhilaf(2, 254, 'warsh'), isFalse);
    });
  });

  // ---- word by word -------------------------------------------------------

  group('word by word', () {
    test("both layers aligned to the ayah's own tokens", () {
      final words = engine.wordByWord.words(112, 1);
      expect(words.map((w) => w.transliteration).toList(),
          ['qul', 'huwa', 'l-lahu', 'aḥadun']);
      expect(words[0].english, 'Say');
      final firstToken =
          engine.quran.ayah(112, 1)!.textArabic.trim().split(_whitespace).first;
      expect(words[0].arabic, firstToken);
    });

    test("every ayah's arrays match its token count", () {
      for (final surah in engine.quran.all()) {
        for (final ayah in surah.ayahs) {
          final tokens = ayah.textArabic.trim().split(_whitespace).length;
          expect(engine.wordByWord.glosses(surah.id, ayah.id)?.length, tokens,
              reason: '${surah.id}:${ayah.id} english');
          expect(engine.wordByWord.transliterations(surah.id, ayah.id)?.length,
              tokens,
              reason: '${surah.id}:${ayah.id} transliteration');
        }
      }
    });

    test('gloss search finds the word', () {
      final hits = engine.wordByWord.find('the Ever-Living', limit: 5);
      expect(hits.any((h) => h.surah == 2 && h.ayah == 255), isTrue);
      expect(hits.every((h) => h.transliteration.isNotEmpty), isTrue);
    });
  });

  // ---- similar ayahs, themes, lessons --------------------------------------

  group('corpora', () {
    test('similar ayahs', () {
      final matches = engine.similarAyahs.matches(2, 255);
      final found = matches.where((m) => m.surah == 3 && m.ayah == 2).toList();
      expect(found, isNotEmpty, reason: '3:2 is a match');
      expect(found.first.verified, isTrue);
      // Version 2 carries no text: the shared wording is a span into the
      // matched ayah.
      expect(found.first.spans, [
        [0, 6]
      ]);
      expect(found.first.labels, isEmpty);
      expect(found.first.score, isNull);
      expect(engine.similarAyahs.has(2, 255), isTrue);
      expect(engine.similarAyahs.count(), greaterThan(5000));
    });

    test('similar ayahs reads only version 2', () {
      // A version-1 file put the phrase text where a span now sits: read as
      // no data, like the JS.
      expect(SimilarAyahs().count(), 0);
      expect(
          SimilarAyahs({
            '2:255': [
              [3, 2, 'text', 1]
            ]
          }).count(),
          0);
      expect(
          SimilarAyahs({
            'v': 1,
            'ayahs': {
              '2:255': [
                [3, 2, 1, [], [], null]
              ]
            }
          }).count(),
          0);
      final two = SimilarAyahs({
        'v': 2,
        'ayahs': {
          '2:255': [
            [
              3,
              2,
              1,
              [
                [0, 6]
              ],
              [],
              null
            ]
          ]
        }
      });
      expect(two.count(), 1);
      expect(two.has(2, 255), isTrue);
      expect(two.has(3, 2), isFalse);
      final match = two.matches(2, 255).single;
      expect(match.surah, 3);
      expect(match.ayah, 2);
      expect(match.verified, isTrue);
      expect(match.labels, isEmpty);
      expect(match.spans, [
        [0, 6]
      ]);
      expect(match.score, isNull);
    });

    test('themes index both ways', () {
      expect(engine.themes.all().length, greaterThanOrEqualTo(300));
      expect(engine.themes.topic('tawheed')!.ayahs, contains('2:255'));
      expect(engine.themes.topicsFor(2, 255).any((t) => t.id == 'tawheed'), isTrue);
      expect(engine.themes.domains().length, greaterThanOrEqualTo(2));
    });

    test('lessons walk in course order', () {
      final lessons = engine.tajweedLessons.allLessons();
      expect(lessons.length, greaterThanOrEqualTo(30));
      expect(engine.tajweedLessons.previous(lessons[0].id), isNull);
      expect(engine.tajweedLessons.next(lessons[0].id)?.id, lessons[1].id);
      expect(engine.tajweedLessons.chapterOf(lessons[0].id), isNotNull);
      // Version 3 examples point at their words by span; the data carries no
      // copy of the words.
      final example = lessons
          .expand((l) => l.examples)
          .firstWhere((e) => e.wordSpan != null);
      expect(example.surahId, 112);
      expect(example.ayahNumber, 1);
      expect(example.wordSpan, [1, 3]);
      expect(const TajweedExample(112, 1, 'focus').wordSpan, isNull);
      expect(
          TajweedExample.fromJson({'surahId': 112, 'ayahNumber': 1, 'focus': ''})
              .wordSpan,
          isNull);
    });

    test("lesson Quran references resolve to the engine's own text", () {
      // Version 4: a drill or rule-card fragment whose Arabic IS Quran carries
      // an `ayah` reference and no text at all, so a port that ignores the
      // field shows an empty row rather than a verse. This reads one back out
      // of the Quran to prove the whole path.
      final lessons = engine.tajweedLessons.allLessons();
      final drill =
          lessons.expand((l) => l.drills).firstWhere((d) => d.ayah != null);
      final reference = drill.ayah!;
      expect(reference.surahId, 110);
      expect(reference.ayahNumber, 1);
      expect([reference.first, reference.last], [0, 4]);
      expect(drill.text, isEmpty,
          reason: 'a referenced drill carries no copy of the words');

      // The span names the whole of an-Nasr 1, so the words it cuts are the
      // ayah itself. Compared against the engine's own text rather than a
      // pasted literal: a copy here would have to be kept in the file's exact
      // normalization, the drift this version removed.
      final words = engine.quran.ayah(110, 1)!.textArabic.split(RegExp(r'\s+'));
      expect(words.length, 5);
      expect(words.sublist(reference.first, reference.last + 1).join(' '),
          words.join(' '));

      // Every reference across drills and rule cards lands inside its ayah.
      var referenced = 0;
      for (final lesson in lessons) {
        final fragments = lesson.ruleCard?.fragments ?? const [];
        for (final row in [...lesson.drills, ...fragments]) {
          final span = row.ayah;
          if (span == null) continue;
          referenced += 1;
          final tokens = engine.quran
              .ayah(span.surahId, span.ayahNumber)!
              .textArabic
              .split(RegExp(r'\s+'));
          expect(span.last, lessThan(tokens.length));
        }
      }
      expect(referenced, 28,
          reason: '7 drills and 21 rule-card fragments reference the Quran');

      // The card itself is `ruleCard`; it was declared as `mushafCard` and so
      // decoded to nothing.
      expect(lessons.where((l) => l.ruleCard != null).length, 32);
    });
  });

  // ---- meaning search ------------------------------------------------------

  test('MaxSim ranks by meaning', () {
    // A toy embedder: enough to prove the scoring, without shipping a model.
    const vectors = <String, List<double>>{
      'patience': [1.0, 0.0, 0.0],
      'hardship': [0.9, 0.1, 0.0],
      'sabr': [0.95, 0.05, 0.0],
      'steadfast': [0.9, 0.05, 0.0],
      'dawn': [0.0, 1.0, 0.0],
      'prayer': [0.0, 0.95, 0.0],
    };
    final semantic = Semantic((word) => vectors[word]).index(const [
      SemanticDocument('sabr', 'sabr and steadfast endurance'),
      SemanticDocument('fajr', 'prayer at dawn'),
    ]);
    final hits = semantic.search('patience hardship');
    expect(hits.first.id, 'sabr');
    expect(hits[0].score, greaterThan(hits[1].score));
  });

  // ---- ask AI --------------------------------------------------------------

  group('ask AI', () {
    test('named verse is the subject', () {
      final passages = engine.askAI.retrieve('explain ayat al-kursi');
      expect(passages.first.reference, '2:255');
      expect(passages.first.isSubject, isTrue);
    });

    test('named surah answers with its background', () {
      final passages = engine.askAI.retrieve('what is surah al-kahf about');
      expect(passages.first.reference, 'Surah Al-Kahf');
      expect(passages.first.kind, PassageKind.surah);
    });

    test('keyword lane is weighted', () {
      final passages =
          engine.askAI.retrieve('what does the Quran say about patience in hardship');
      expect(passages.any((p) => p.reference == '2:153'), isTrue);
      for (final passage in passages) {
        if (passage.kind == PassageKind.ayah) {
          expect(engine.quran.ayah(passage.surah!, passage.ayah!), isNotNull);
        }
      }
    });

    test('bare follow-up uses the previous question', () {
      final carried = engine.askAI.retrieve('tell me about 2:153');
      expect(engine.askAI.retrieve('why?'), isEmpty);
      final withContext = engine.askAI.retrieve(
        'why?',
        previousQuestion: 'tell me about 2:153',
        carried: carried,
      );
      expect(withContext.any((p) => p.reference == '2:153'), isTrue);
    });

    test('prompt shape', () {
      final passages = engine.askAI.retrieve('explain 2:153');
      final built = chatPrompt('explain 2:153', passages);
      expect(built.instructions, contains('Never issue a religious ruling'));
      expect(built.prompt, contains('SUBJECT OF THE QUESTION [2:153]'));
      expect(built.prompt.trimRight(), endsWith('QUESTION: explain 2:153'));
      expect(questionWords, contains('what'));
    });
  });

  // ---- surah sections -----------------------------------------------------

  group('surah sections', () {
    test('111 surahs carry an outline, and it reads as a chain', () {
      expect(engine.surahSections.count(), 111);
      expect(engine.surahSections.overview(1).length, greaterThan(20));
      expect(engine.surahSections.hasSections(1), isFalse);

      // Hud opens with a broad passage and the sections inside it - an ayah has
      // a chain, not a row.
      expect(engine.surahSections.sectionsFor(11, 3).map((s) => s.english).toList(),
          ['Doctrine facts', 'Calling to Allah']);
      expect(engine.surahSections.sectionFor(11, 3)?.english, 'Calling to Allah');
    });

    test('the outline rebuilds the nesting the flat list encodes', () {
      final roots = engine.surahSections.outline(11);
      expect(roots.length, greaterThanOrEqualTo(2));
      expect(roots[0].section.english, 'Doctrine facts');
      expect(roots[0].children.length, 5);
      expect(
        roots[0].children.every((c) =>
            c.section.from >= roots[0].section.from &&
            c.section.to <= roots[0].section.to),
        isTrue,
      );
    });

    test('every range is inside its surah, and search finds a story', () {
      for (final surah in engine.quran.all()) {
        for (final section in engine.surahSections.sections(surah.id)) {
          expect(section.from, greaterThanOrEqualTo(1));
          expect(section.to, lessThanOrEqualTo(surah.numberOfAyahs));
          expect(section.from, lessThanOrEqualTo(section.to));
          expect(section.english, isNotEmpty);
          expect(section.arabic, isNotEmpty);
        }
      }
      final nuh = engine.surahSections.search('Story of Nuh');
      expect(nuh.length, greaterThanOrEqualTo(2));
      expect(nuh.any((hit) => hit.surah == 11), isTrue);
    });
  });

  // ---- arabic alphabet ----------------------------------------------------

  group('arabic alphabet', () {
    test('28 letters, each with a tajweed weight the catalogue explains', () {
      final letters = engine.alphabet.letters();
      expect(letters.length, 28);
      final weights = engine.alphabet.weightDescriptions();
      for (final letter in letters) {
        expect(letter.letter, isNotEmpty);
        expect(letter.name, isNotEmpty);
        expect(letter.weight, isNotNull, reason: '${letter.letter} has no weight');
        expect(weights[letter.weight], isNotNull,
            reason: 'no description for weight ${letter.weight}');
      }
      // Alif is the one letter with no weight of its own - the fact the whole
      // field exists for.
      expect(engine.alphabet.weight('ا'), 'followsPrevious');
      expect(engine.alphabet.weight('ص'), 'heavy');
      expect(engine.alphabet.weight('س'), 'light');
    });

    test('a letter resolves from any of its joining forms', () {
      expect(engine.alphabet.letter('ـصـ')?.transliteration, 'Saad');
      expect(engine.alphabet.letterById(1)?.letter, 'ا');
      expect(engine.alphabet.letter('nope'), isNull);
    });

    test('tashkeel, numerals and the waqf signs', () {
      expect(engine.alphabet.tashkeel().length, greaterThanOrEqualTo(8));
      expect(engine.alphabet.numbers().length, 11);
      expect(engine.alphabet.stoppingSign('۩')?.title, 'Make Sujood');
      expect(engine.alphabet.heavyLetters().length, greaterThanOrEqualTo(7));
    });
  });

  // ---- qiraat comparison --------------------------------------------------

  group('qiraat comparison', () {
    test('only the published readings, hafs included', () {
      expect(engine.qiraatComparison.available(),
          ['buzzi', 'duri', 'hafs', 'qaloon', 'qunbul', 'shubah', 'susi', 'warsh']);
    });

    test('a reading against itself is entirely identical', () {
      final same = engine.qiraatComparison.compareSurah(2, 'hafs');
      expect(same.identical, same.words);
      expect(same.sameSkeleton, 0);
      expect(same.different, 0);
      expect(same.added, 0);
      expect(same.dropped, 0);
      expect(same.identicalPercent, 100);
    });

    test('the buckets add up, and Shubah is nearer to Hafs than Warsh is', () {
      final warsh = engine.qiraatComparison.compareSurah(2, 'warsh');
      final shubah = engine.qiraatComparison.compareSurah(2, 'shubah');
      for (final totals in [warsh, shubah]) {
        expect(
          totals.identical + totals.sameSkeleton + totals.different + totals.dropped,
          totals.words,
        );
      }
      // Shu'bah and Hafs are the two narrators of Asim; Warsh is a different imam.
      expect(shubah.identicalPercent, greaterThan(warsh.identicalPercent));
      expect(shubah.identicalPercent, greaterThan(95));
    });

    test('differences are the non-identical rows, in reading order', () {
      final rows = engine.qiraatComparison.differences(2, 'warsh', limit: 20);
      expect(rows, isNotEmpty);
      expect(rows.every((row) => row.base != row.other), isTrue);
      final positions = rows.map((row) => row.position).toList();
      expect(positions, orderedEquals(positions.toList()..sort()));
    });

    test("word streams follow the reading's own verse count", () {
      // Warsh reads al-Baqarah as 285 verses to Hafs' 286, so pairing by ayah id
      // would compare different verses; the comparison walks the words instead.
      expect(engine.quran.numberOfAyahsInQiraah(2, 'warsh'), 285);
      expect(engine.quran.qiraahVerses(2, 'warsh').length, 285);
      expect(engine.qiraatComparison.words(2, 'warsh').length, greaterThan(6000));
    });
  });
}
